### DISCUSSION TITLE
IEEE-754 compliant arithmetic

### DISCUSSION THREAD

**Discussion by soerenwolfers** [duckdb/duckdb#10956] [General]
I think returning IEEE-754 answers for division by zero would be another great way in which duckdb could be more friendly and modern than old and dusty alternative databases. 

Currently, duckdb is compliant on infinities but not division by zeros, which gets it shared second rank among the databases I tested

|            | 1.0 / 0.0   |  0.0 / 0.0   | -1.0 / 0.0   | inf/inf   | 1.0 / inf   | inf-inf   |   inf - 1.0 |
|:-----------|:------|:------|:-------|:----------|:--------|:----------|----------:|
| IEEE       | inf   | nan   | -inf   | nan       | 0       | nan       |       inf |
| duckdb     | NULL  | NULL  | NULL   | -nan      | 0       | -nan      |       inf |
| postgresql | err   | err   | err    | nan       | 0       | nan       |      -inf |
| clickhouse | inf   | nan   | -inf   | nan       | 0       | nan       |       inf |
| mysql      | NULL  | NULL  | NULL   | NULL      | NULL    | 0         |         1 |
| sqlite     | NULL  | NULL  | NULL   | NULL      | NULL    | 0         |         1 |

Related, I think `sqrt` / `sin` / `asin` / ... should return `NaN` as well where they currently return `OutOfRange` or `InvalidInput` errors (e.g. on negative / infinite / larger than 1 inputs). 

Tickets opened after this discussion that are asking for essentially the same (updating this list by hand whenever I see a new one):

- https://github.com/duckdb/duckdb/issues/11374
-  https://github.com/duckdb/duckdb/issues/11450
- https://github.com/duckdb/duckdb/issues/11442
- https://github.com/duckdb/duckdb/issues/14373
- https://github.com/duckdb/duckdb/issues/14249
- https://github.com/duckdb/duckdb/discussions/17341

**Comment by szarnyasg** [comment]
@soerenwolfers I added a section about this to the Postgres compatibility page: https://github.com/duckdb/duckdb-web/pull/3185

I got slightly different results from your table though, e.g., I never got a `-NaN` value.

**Reply by soerenwolfers** [reply]
IEEE754 doesn't specify the behavior of the sign bit of NaNs, so that's fine. 

By the way: thanks for adding that section, but I, and most other duckdb users I know, haven't used PostgreSQL or any other database query language before and would therefore not consider reading the PostgreSQL compatibility page (and from what I can tell, people that **do** have a databases background don't typically expect much sanity from floating point operations to begin with). That's more of a general remark though and having an intro to duckdb for people with, say, a data science or numerical simulation background is maybe a bigger and unrelated project.

**Reply by szarnyasg** [reply]
Good point, I'll also add this to the numeric page.

**Comment by dmcmurchy** [comment]
Slightly related is that if you try to run `Summarize` on a table that contains `NaN` values it also throws: `OutOfRangeException: Out of Range Error: STDDEV_SAMP is out of range!` 

Took me some time to figure out why as I didn't realize that DuckDB wasn't treating NaN values as NULL which I was used to with Pandas. 

Another thing that tripped me up was the output in VSCode showing NULLS as NaN when displayed as a DataFrame (lowest output). 

![image](https://github.com/duckdb/duckdb/assets/5430744/239d38be-adfe-4288-b210-336a3f5ff617)

**Reply by soerenwolfers** [reply]
Treating NaN as NULL is the opposite of IEEE compliance. Nevertheless, I agree that duckdb should return NaN instead of OutOfRange errors.

VScode isn't showing NULLs as NaNs; it's displaying a pandas DataFrame and duckdb converts NULLs to NaNs on conversion to pandas Dataframes. That being said, I do think it should use nullable columns instead, and recommend you open a separate ticket about that (a related ticket, which complains that the inverse pandas->duckdb conversion converts NaN to NULL, was previously closed as won't fix, so I don't have a lot of hopes but it can't harm if you ask).

**Comment by dmcmurchy** [comment]
This might be of use to people that come across the topic to show what DuckDB does as of version 1.2.1 when it comes to dealing with divide by zero, infinity, null & nan.

```sql
FROM (
    SELECT 
        /* Division by zero */
        "1 / 0": 1 / 0,
        "1.0 / 0.0": 1.0 / 0.0, 
        "-1.0 / 0.0": -1.0 / 0.0,
        "0 / 0": 0 / 0,
        "0.0 / 0.0": 0.0 / 0.0,
        "nan / 0": 'Nan'::FLOAT / 0, -- cast required
        "NULL / 0": NULL / 0,
        "Infinity / 0": 'Infinity'::FLOAT / 0, -- cast required
        
        /* Division by NULL */
        "0 / NULL": 0.0 / NULL,
        "1 / NULL": 1 / NULL,
        "NULL / NULL": NULL / NULL,
        "nan / NULL": 'Nan'::FLOAT / NULL, -- cast required
        "Infinity / NULL": 'Infinity'::FLOAT / NULL, -- cast required
        
        /* Division by Infinity */
        "nan / Infinty": 'Nan'::FLOAT / 'Infinity'::FLOAT, -- cast required
        "Infinity / Infinty": 'Infinity'::FLOAT / 'Infinity'::FLOAT, -- cast required
        "1.0 / Infinity": 1.0 / 'Infinity'::FLOAT, -- cast required
        
        /* Division by nan */
        "Infinity / nan": 'Infinity'::FLOAT / 'nan'::FLOAT, -- cast required
        "0 / nan":  0 / 'nan', 
        "nan / nan":  'nan'::FLOAT / 'nan'::FLOAT, -- cast required

        "1 / 2": 1 / 2,
        "1.0 / 2.0": 1.0 / 2.0,

        /* Subtracting Infinity or Nan*/
        "Infinity - Infinity": 'Infinity'::FLOAT - 'Infinity'::FLOAT, -- cast required
        "nan - nan": 'nan'::FLOAT - 'nan'::FLOAT, --cast required
        
)
UNPIVOT INCLUDE NULLS -- required to show nulls in output
(
    result FOR expression IN (
        COLUMNS(*)
    )
)
LIMIT 25;
```
| expression (varchar)         |   result (double) |
|:--------------------|---------:|
| 1 / 0               |    inf   |
| 1.0 / 0.0           |    inf   |
| -1.0 / 0.0          |   -inf   |
| 0 / 0               |    -nan   |
| 0.0 / 0.0           |    -nan   |
| nan / 0             |    nan   |
| NULL / 0            |   NULL   |
| Infinity / 0        |    inf   |
| 0 / NULL            |    NULL   |
| 1 / NULL            |    NULL   |
| NULL / NULL         |    NULL   |
| nan / NULL          |    NULL   |
| Infinity / NULL     |    NULL   |
| nan / Infinty       |    nan   |
| Infinity / Infinty  |    -nan   |
| 1.0 / Infinity      |      0.0   |
| Infinity / nan      |    nan   |
| 0 / nan             |    nan   |
| nan / nan           |    nan   |
| 1 / 2               |      0.5 |
| 1.0 / 2.0           |      0.5 |
| Infinity - Infinity |    -nan   |
| nan - nan           |    nan   |

**Reply by soerenwolfers** [reply]
Note that this can be summarized as "DuckDB arithmetic (but not yet the special functions) is now IEEE754 compliant for non-NULL values (i.e. behaves the same  as most other programming languages out there do on capable hardware) and returns NULL whenever any operand is NULL (i.e. propagates NULLs like most other DuckDB operations)."
